Source/Evolve/WarpXComputeDt.cpp

```diff
@@ -0,0 +1,85 @@
+/* Copyright 2020
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "WarpX.H"
+#include "Utils/WarpXAlgorithmSelection.H"
+#ifdef WARPX_DIM_RZ
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H"
+#else
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianYeeAlgorithm.H"
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianCKCAlgorithm.H"
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianNodalAlgorithm.H"
+#endif
+
+void
+WarpX::ComputeDt ()
+{
+    // Determine
+    const amrex::Real* dx = geom[max_level].CellSize();
+    amrex::Real deltat = 0.;
+
+#if WARPX_USE_PSATD
+    // Computation of dt for spectral algorithm
+
+#   if (defined WARPX_DIM_RZ)
+    // - In RZ geometry: dz/c
+    deltat = cfl * dx[1]/PhysConst::c;
+#   elif (defined WARPX_DIM_XZ)
+    // - In Cartesian 2D geometry: determined by the minimum cell size in all direction
+    deltat = cfl * std::min( dx[0], dx[1] )/PhysConst::c;
+#   else
+    // - In Cartesian 3D geometry: determined by the minimum cell size in all direction
+    deltat = cfl * std::min( dx[0], std::min( dx[1], dx[2] ) )/PhysConst::c;
+#   endif
+
+
+#else
+    // Computation of dt for FDTD algorithm
+
+#   ifdef WARPX_DIM_RZ
+    // - In RZ geometry
+    if (maxwell_fdtd_solver_id == MaxwellSolverAlgo::Yee) {
+        deltat = cfl * CylindricalYeeAlgorithm::ComputeMaxDt(dx,  n_rz_azimuthal_modes);
+#   else
+    // - In Cartesian geometry
+    if (do_nodal) {
+        deltat = cfl * CartesianNodalAlgorithm::ComputeMaxDt( dx );
+    } else if (maxwell_fdtd_solver_id == MaxwellSolverAlgo::Yee) {
+        deltat = cfl * CartesianYeeAlgorithm::ComputeMaxDt( dx );
+    } else if (maxwell_fdtd_solver_id == MaxwellSolverAlgo::CKC) {
+        deltat = cfl * CartesianCKCAlgorithm::ComputeMaxDt( dx );
+#   endif
+    } else {
+        amrex::Abort("Unknown algorithm");
+    }
+
+#endif
+
+    dt.resize(0);
```

Comment:
Is there a reason that dt is resized twice?

Reply:
I think that the first resize is to make sure that there is no previously-existing value in `dt` ; otherwise the second `resize` will preserve these pre-existing values instead of overwriting them. (Note that I mindlessly copied this part from the previously-existing code that calculated `dt` before this PR.)
By agree that this looks confusing ; I'll update the PR and add a comment.

Reply:
Ok. It doesn't seem like it is needed, but it's probably safer to keep it rather than risk breaking things. An alternative is to do `dt.clear()` which will do the same thing but makes its intent clear. Maybe @WeiqunZhang knows why it is there (since he wrote it).